src/optimizers/quantizer.cpp

```diff
@@ -0,0 +1,187 @@
+ * @param size the data size
+ * @param numCenters the number of quantized centers in absolute. It should be 2^(bit-1)
+ * @param S stores the scaling factor.
+ * @param base for log quantized center. Default of 2
+ */
+void quantizeLog(Tensor data, Tensor res, int numCenters, float S, float base = 2.0f) {
+  using namespace functional;
+
+  // clip based on the scaling factor
+  Element(_1 = clip(_2, S), res, data);
+
+  // multiplier such that the quantization is rounded in normal-space instead of log space.
+  // 4/3 for base = 2. example: 11.8 should be quantized to 8, instead of 16.
+  float mult = (2.0f * base) / (1.0f + base);
+
+  // log-quantization works as the following:
+  // 1. capture the sign:
+  // sign = sgn(v)
+  // 2. get the quantization center:
+  // qc = floor(log2(abs(v/S) * _mult))
+  // 3. clip the center to make sure we have no more than 2^(bit-1) centers.
+  // qc = clip(qc, num_centers)
+  // 4. revert back to floating point space:
+  // q = 2^qc * S * sign
+  //
+  // The above steps are writen in 1 call as below, to avoid reserving extra Tensors:
+
+  Element(
+      _1 = sgn(_1)      // revert the sign back
+           * S          // revert the scaling function
+           * pow(base,  // revert from log space to normal FP represtation
+                 clip(floor(log(abs(_1 / S) * mult) / log(base)),  // get the quantization center
+                      (float) numCenters)),                                // clip
+      res);
+}
+
+/* Quantize all the parameters (except bias, unless enabled via --quantize-bias)
+ * Quantization only works if we store the quantization error residual
+ * The stored residual will be added for the next quantization
+ * @param graph is the model graph to be quantized (in-place) 
+ */
+void ModelQuantizer::quantize(Ptr<ExpressionGraph> graph) {
+  // reserve tensor for error feedback mechanism
+  if(!errorResidual_) {
+    LOG(info, " EXPERIMENTAL: quantize the model to {}-bits", bits_);
+
+    int numElements = (int)graph->params()->vals()->size();
+    auto allocator = New<TensorAllocator>(graph->getBackend());
+    allocator->reserveExact(graph->params()->vals()->memory()->size());
+    allocator->allocate(errorResidual_, {1, numElements});
+
+    allocators_.push_back(allocator);
+    firstError_ = true;
+  }
+
+  {
+    // apply error feedback mechanism
+    using namespace functional;
+    // add the previous error residual to the current model
+    Element(_1 += _2, graph->params()->vals(), errorResidual_);
+    errorResidual_->copyFrom(graph->params()->vals());
+  }
+
+  for(auto p : *graph->params()) {
+    // optionally quantize bias
+    if(quantBias_ || p->val()->shape()[0] > 1)
+      quantizeImpl(p->val());
+  }
+
+  // get new error residual. Skip the first one.
+  if (!firstError_) {
+    using namespace functional;
+    Element(_1 -= _2, errorResidual_, graph->params()->vals());
+  }
+  else {
+    errorResidual_->set(0);
```

Comment:
Not critical, but do we really save so much by skipping the first one? Can we not just initialize it to 0 when we allocate the error residual, and remove this condition here?

Reply:
We're not sure if storing the first error yields a better result. The original method is to skip the first quantization error, so we'll stick to the default. However, I am experimenting if we can just remove this condition.